File: app/src/main/java/com/loony/timelapsemaker/dialog_settings/DialogSettingsAdapter.java
package com.loony.timelapsemaker.dialog_settings;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.SwitchCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.loony.timelapsemaker.R;
import com.loony.timelapsemaker.Util;

import java.util.ArrayList;

/**
 * Created by Kamil on 7/25/2017.
 */

public class DialogSettingsAdapter extends ArrayAdapter<DialogOption> {
    private Context context;
    private LayoutInflater inflater;
    private ArrayList<DialogOption> options;

    public DialogSettingsAdapter(Context context, ArrayList<DialogOption> options) {
        super(context, 0, options);
        this.context = context;
        this.options = options;
        this.inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        Util.log("Is this visible that calling?");
        Holder holder;
        View view;

        if (convertView == null) {
            view = inflater.inflate(R.layout.item_settings_element, parent, false);
            holder = new Holder();
            holder.icon = (ImageView) view.findViewById(R.id.icon);
            holder.label = (TextView) view.findViewById(R.id.label);
            holder.description = (TextView) view.findViewById(R.id.description);
            holder.switchCompat = (SwitchCompat) view.findViewById(R.id.switchBtn);

            view.setTag(holder);
        } else {
            view = convertView;
            holder = (Holder) view.getTag();
        }

        DialogOption item = getItem(position);
        holder.icon.setImageResource(item.icon);
        holder.label.setText(item.label);
        holder.description.setText(item.description);

        if(item.switchState != DialogOption.Switch.NOT_EXIST) {
            holder.switchCompat.setVisibility(View.VISIBLE);
            holder.switchCompat.setChecked(item.switchState == DialogOption.Switch.ENABLED);
        } else {
            holder.switchCompat.setVisibility(View.GONE);
        }

        return view;
    }

    private class Holder {
        public ImageView icon;
        public TextView label, description;
        public SwitchCompat switchCompat;
    }
}

File: app/src/main/java/pl/kflorczyk/timelapsemaker/dialog_settings/DialogSettingsAdapter.kt
package pl.kflorczyk.timelapsemaker.dialog_settings

import android.content.Context
import android.support.v7.widget.SwitchCompat
import android.widget.TextView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import android.widget.ArrayAdapter
import android.widget.ImageView
import pl.kflorczyk.timelapsemaker.R

class DialogSettingsAdapter(context: Context, options: ArrayList<DialogOption>) : ArrayAdapter<DialogOption>(context, 0, options) {
    private val inflater: LayoutInflater = LayoutInflater.from(context)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: Holder
        val view: View

        if (convertView == null) {
            view = inflater.inflate(R.layout.item_settings_element, parent, false)
            holder = Holder()
            holder.icon = view.findViewById(R.id.icon) as ImageView
            holder.label = view.findViewById(R.id.label) as TextView
            holder.description = view.findViewById(R.id.description) as TextView
            holder.switchCompat = view.findViewById(R.id.switchBtn) as SwitchCompat

            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as Holder
        }

        val item = getItem(position)
        holder.icon!!.setImageResource(item!!.icon)
        holder.label!!.text = item.label
        holder.description!!.text = item.description

        if (item.switchState !== DialogOption.Switch.NOT_EXIST) {
            holder.switchCompat!!.visibility = View.VISIBLE
            holder.switchCompat!!.isChecked = item.switchState === DialogOption.Switch.ENABLED
            holder.switchCompat!!.setOnCheckedChangeListener(item.onSwitchCheckedChange)
        } else {
            holder.switchCompat!!.visibility = View.GONE
        }

        return view
    }

    private inner class Holder {
        var icon: ImageView? = null
        var label: TextView? = null
        var description: TextView? = null
        var switchCompat: SwitchCompat? = null
    }
}